### ISSUE TITLE
Files in sqlfluffignore are being linted

### ISSUE THREAD

**Issue by jelstongreen** [sqlfluff/sqlfluff#2545]
<!--If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant.-->

Files containing macros that seem to be unsupported by sqlfluff (another issue I have not raised) are places in the .sqlfluffignore file but are still being linted. The files have dbt compile error but compile fine with `dbt run`

## Expected Behaviour
Files should be ignored
## Observed Behaviour
Files are run with templating error
## Steps to Reproduce
```
➜  datalake-models git:(lint-core) ✗ sqlfluff lint core/models                                                   
=== [dbt templater] Sorting Nodes...                                                                                                                                                                                                                                                                                    
=== [dbt templater] Compiling dbt project...                                                                                                                                                                                                                                                                            
=== [dbt templater] Project Compiled.                                                                                                                                                                                                                                                                                   
== [core/models/model_a.sql] FAIL                                                                                                                                                                                                                                                
L:   0 | P:   0 |  TMP | dbt compilation error on file                                                                                                                                                                                                                                                                  
                       | 'core/models/model_a.sql',
                       | 'dict object' has no attribute 'nodes'
== [core/models/model_b.sql] FAIL                                                                                                                                                                                                                                                     
L:   0 | P:   0 |  TMP | dbt compilation error on file
                       | 'core/models/model_b.sql',
                       | 'dict object' has no attribute 'nodes'
All Finished 📜 🎉!   
```
.sqlfluffignore
```
model_a.sql
model_b.sql
```
OR
```
core/models/model_a.sql
core/models/model_b.sql
```

## Dialect
dbt spark (databricks)

## Version
<!-- Include the output of `sqlfluff --version` along with your Python version.
If you are using dbt then please additionally include the dbt version. -->
sqlfluff, version 0.9.4
dbt --version                                                                
installed version: 1.0.1
   latest version: 1.0.1

Up to date!

Plugins:
  - snowflake: 0.21.1
  - redshift: 0.21.1
  - postgres: 0.21.1
  - spark: 1.0.0
                     
Python 3.9.10

## Configuration
<!--
Include your SQLFluff configuration here
-->
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = spark3
templater = dbt
rules = None
exclude_rules = L031
recurse = 0
output_line_length = 80
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
disable_noqa = False
# Comma separated list of file extensions to lint.
# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:indentation]
indented_joins = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:dbt]
apply_dbt_builtins = True
load_macros_from_path=core/macros

[sqlfluff:templater:dbt:context]
DATABRICKS_TARGET=xxx
tbl_name=my_table

[sqlfluff:templater:dbt:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 2
max_line_length = 99
indent_unit = space
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

[sqlfluff:rules:L011]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L012]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = consistent

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L019]
comma_style = leading

[sqlfluff:rules:L026]
force_enable = False

[sqlfluff:rules:L028]
force_enable = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = upper


[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = upper

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L052]  # Semi-colon formatting approach.
multiline_newline = False
require_final_semicolon = False

[sqlfluff:rules:L054]  # GROUP BY/ORDER BY column references.
group_by_and_order_by_style = explicit

[sqlfluff:rules:L057]  # Special characters in identifiers
unquoted_identifiers_policy = all
quoted_identifiers_policy = all
allow_space_in_identifier = False
```

**Comment by barrywhart** [comment]
@jelstongreen: We need more info before we can investigate this issue. Can you provide an example of the failing files? It's _possible_ that dbt itself is reading the files and having issues, not SQLFluff.

BTW, we recently added support for files containing macro definitions. See issue #2456. This fix is not yet released, but you may want to test it and see if it fixes your issue. We will likely make a new SQLFluff release this week.

**Comment by barrywhart** [comment]
@jelstongreen: I'm getting an error from `dbt compile`:
```
root@42e8a47f04f3:/app/my-dbt# dbt compile --models models/model_a.sql 
12:39:19  Running with dbt=1.0.1
12:39:20  Unable to do partial parsing because profile has changed
12:39:27  Encountered an error:
Compilation Error in model model_a (models/model_a.sql)
  'project_z' is undefined
  
  > in macro hash_sensitive_columns (macros/hash_sensitive_columns.sql)
  > called by model model_a (models/model_a.sql)
```

**Comment by jelstongreen** [comment]
That's just where I have renamed our actual project. You can remove all the project_z prefixes I think.

**Comment by barrywhart** [comment]
@jelstongreen: I don't really know dbt, so I need a good example to work from. Can you update your example, please? I'm having trouble making the changes you suggest. Thanks!

**Comment by jelstongreen** [comment]
Due to the complexity around the macro set up I think the easiest way of showing this to you was to create a repo so I have forked jaffle_shop and reformatted to show the example:

https://github.com/jelstongreen/jaffle_shop

 The only thing you should need to do it is make sure you have a `SALT` env variable:
`export SALT=SALTY`
and make sure the profiles.yml entry for schema is `jaffle_shop` or you can change the variables in the `hash_sensitive_columns` and `get_meta_columns` macros that say `project='jaffle_shop'` to be `project='<your_schema>`

**Comment by barrywhart** [comment]
Thanks, I'll take a look.

Yes, dbt is tricky to support: complex projects and files = tricky repro scenarios.

**Comment by barrywhart** [comment]
@jelstongreen: I made some progress, but I'm still getting errors. Perhaps I need to create some tables in my database? See the `WARNING` messages below, then the final error:
```
18:19:14  Compilation Error in model stg_customer_snapshot (models/staging/stg_customer_snapshot.sql)
18:19:14    'dict object' has no attribute 'model.default.stg_customer_snapshot_base'
```

Full output from `dbt build`:
```
root@42e8a47f04f3:/app/issue_2545# dbt build -s +stg_customer_snapshot
18:19:03  Running with dbt=1.0.1
18:19:03  Unable to do partial parsing because profile has changed
18:19:10  [WARNING]: Did not find matching node for patch with name 'stg_customers' in the 'models' section of file 'models/staging/schema.yml'
18:19:10  [WARNING]: Did not find matching node for patch with name 'stg_customer' in the 'models' section of file 'models/staging/schema.yml'
18:19:10  [WARNING]: Test 'test.jaffle_shop.unique_stg_customers_customer_id.c7614daada' (models/staging/schema.yml) depends on a node named 'stg_customers' which was not found
18:19:10  [WARNING]: Test 'test.jaffle_shop.not_null_stg_customers_customer_id.e2cfb1f9aa' (models/staging/schema.yml) depends on a node named 'stg_customers' which was not found
18:19:10  [WARNING]: Test 'test.jaffle_shop.unique_stg_customer_id.359a723c10' (models/staging/schema.yml) depends on a node named 'stg_customer' which was not found
18:19:10  [WARNING]: Test 'test.jaffle_shop.not_null_stg_customer_id.1fe2201576' (models/staging/schema.yml) depends on a node named 'stg_customer' which was not found
18:19:10  Found 2 models, 1 test, 1 snapshot, 0 analyses, 356 macros, 0 operations, 1 seed file, 0 sources, 0 exposures, 0 metrics
18:19:10  
18:19:11  Concurrency: 4 threads (target='dev')
18:19:11  
18:19:11  1 of 5 START seed file dbt_alice.raw_customers.................................. [RUN]
18:19:12  1 of 5 OK loaded seed file dbt_alice.raw_customers.............................. [INSERT 100 in 0.92s]
18:19:12  2 of 5 START snapshot src_jaffle_shop_snapshots.customers_snapshot.............. [RUN]
18:19:13  2 of 5 OK snapshotted src_jaffle_shop_snapshots.customers_snapshot.............. [SELECT 100 in 0.67s]
18:19:13  3 of 5 START view model dbt_alice.stg_customer_snapshot_base.................... [RUN]
18:19:13  3 of 5 OK created view model dbt_alice.stg_customer_snapshot_base............... [CREATE VIEW in 0.42s]
18:19:13  4 of 5 START test not_null_stg_customer_snapshot_base_id........................ [RUN]
18:19:14  4 of 5 PASS not_null_stg_customer_snapshot_base_id.............................. [PASS in 0.34s]
18:19:14  5 of 5 START view model dbt_alice.stg_customer_snapshot......................... [RUN]
18:19:14  5 of 5 ERROR creating view model dbt_alice.stg_customer_snapshot................ [ERROR in 0.04s]
18:19:14  
18:19:14  Finished running 1 seed, 1 snapshot, 2 view models, 1 test in 3.55s.
18:19:14  
18:19:14  Completed with 1 error and 0 warnings:
18:19:14  
18:19:14  Compilation Error in model stg_customer_snapshot (models/staging/stg_customer_snapshot.sql)
18:19:14    'dict object' has no attribute 'model.default.stg_customer_snapshot_base'
18:19:14    
18:19:14    > in macro hash_sensitive_columns (macros/hash_sensitive_columns.sql)
18:19:14    > called by macro get_meta_columns (macros/get_meta_columns.sql)
18:19:14    > called by model stg_customer_snapshot (models/staging/stg_customer_snapshot.sql)
18:19:14  
18:19:14  Done. PASS=4 WARN=0 ERROR=1 SKIP=0 TOTAL=5
```

**Comment by jelstongreen** [comment]
Is your schema in your profiles.yml set to `default`?

**Comment by barrywhart** [comment]
This is my profiles.yml:
```
root@42e8a47f04f3:/app/issue_2545# cat ~/.dbt/profiles.yml 
default:
  target: default
  outputs:
    default:
      type: postgres
      host: postgres
      user: postgres
      pass: password
      port: 5432
      dbname: postgres
      schema: default
      threads: 4

config:
  send_anonymous_usage_stats: false
```

**Comment by barrywhart** [comment]
I never got `dbt build` to work, but I did get `dbt compile` to work, after making a few code changes. I think `dbt build` is trying to actually run the SQL, and it only works if certain tables are present.

I also ran `sqlfluff lint`. If my `.sqlfluffignore` file includes the line `stg_customer_snapshot.sql`, then I don't get any lint warnings for that file. If I remove that line, then I do. Does this info help?

To be clear, this is the contents of `.sqlfluffignore`:
```
target/
dbt_packages/
macros/
#jaffle_shop/models/staging/stg_customer_snapshot.sql
stg_customer_snapshot.sql
```

**Comment by jelstongreen** [comment]
Ah ok interesting, so it will not find the file if the path is included, it has to be just the model name plus .sql file extension. Great, one problem sorted! Any thoughts on why dbt compile is working but sqlfluff cannot parse the file?

**Comment by barrywhart** [comment]
I'm not certain of the "rules" for sqlfluffignore, just wanted you to know that I got something working. 

I don't know about the other question, because I couldn't reproduce it. Based on the limited information you've provided (the error message), I wouldn't call it a parsing issue. It seems more likely to be a failure in some of the functions or macros called in the template. If you can get us a better repro scenario, we can take another look.

Do you want to try and do that here, or shall we close this issue?

**Comment by jelstongreen** [comment]
So if you remove `stg_customer_snapshot.sql` and then `sqlfluff lint`, do you get an error on that file?

**Comment by jelstongreen** [comment]
The only table inputs in that repo come from a seed `raw_customers.csv`, nothing is pre-created in an underlying database that I have access too and you don't so it should build the data lineage the same way in theory. Could you share your latest error from `dbt build` now that you have `dbt compile` working?

**Comment by jelstongreen** [comment]
Would also be good to have a look at your `target/compiled/jaffle_shop/models/staging/stg_customer_snapshot.sql` if you are able to compile it too 🙏🏼

**Comment by barrywhart** [comment]
This is the output I get from `dbt build`:
```
root@42e8a47f04f3:/app/issue_2545# dbt build -s +stg_customer_snapshot
14:46:27  Running with dbt=1.0.1
14:46:28  Found 2 models, 1 test, 1 snapshot, 0 analyses, 356 macros, 0 operations, 1 seed file, 0 sources, 0 exposures, 0 metrics
14:46:28  
14:46:28  Concurrency: 4 threads (target='default')
14:46:28  
14:46:28  1 of 5 START seed file default.raw_customers.................................... [RUN]
14:46:29  1 of 5 OK loaded seed file default.raw_customers................................ [INSERT 100 in 0.82s]
14:46:29  2 of 5 START snapshot src_jaffle_shop_snapshots.customers_snapshot.............. [RUN]
14:46:30  2 of 5 OK snapshotted src_jaffle_shop_snapshots.customers_snapshot.............. [INSERT 0 0 in 1.07s]
14:46:30  3 of 5 START view model default.stg_customer_snapshot_base...................... [RUN]
14:46:31  3 of 5 OK created view model default.stg_customer_snapshot_base................. [CREATE VIEW in 0.41s]
14:46:31  4 of 5 START test not_null_stg_customer_snapshot_base_id........................ [RUN]
14:46:31  4 of 5 PASS not_null_stg_customer_snapshot_base_id.............................. [PASS in 0.32s]
14:46:31  5 of 5 START view model default.stg_customer_snapshot........................... [RUN]
14:46:31  5 of 5 ERROR creating view model default.stg_customer_snapshot.................. [ERROR in 0.04s]
14:46:31  
14:46:31  Finished running 1 seed, 1 snapshot, 2 view models, 1 test in 3.51s.
14:46:31  
14:46:31  Completed with 1 error and 0 warnings:
14:46:31  
14:46:31  Compilation Error in model stg_customer_snapshot (models/staging/stg_customer_snapshot.sql)
14:46:31    'dict object' has no attribute 'model.default.stg_customer_snapshot_base'
14:46:31    
14:46:31    > in macro hash_sensitive_columns (macros/hash_sensitive_columns.sql)
14:46:31    > called by macro get_meta_columns (macros/get_meta_columns.sql)
14:46:31    > called by model stg_customer_snapshot (models/staging/stg_customer_snapshot.sql)
14:46:31  
14:46:31  Done. PASS=4 WARN=0 ERROR=1 SKIP=0 TOTAL=5
```

**Comment by barrywhart** [comment]
And this is my compiled `stg_customer_snapshot.sql`. Note that I had hacked one of the macro files to hardcode a list of columns, c1, c2, because the actual table didn't exist.

```
WITH stg_customer AS (
    SELECT
      

    
        

    SHA2(
        TRIM(
            LOWER(
                CAST(c1 AS 
    varchar
)
                || 'SALTY'
            )
        ),
        256
    ) AS c1_hash,


    
        

    SHA2(
        TRIM(
            LOWER(
                CAST(c2 AS 
    varchar
)
                || 'SALTY'
            )
        ),
        256
    ) AS c2_hash,


    

    "customer_snapshot_id" as "customer_snapshot_id",
  "id" as "id",
  "first_name" as "first_name",
  "last_name" as "last_name"


    FROM "postgres"."default"."stg_customer_snapshot_base"
)

SELECT *
FROM stg_customer
```

**Comment by jelstongreen** [comment]
Ok strange, must be due to the macro being incompatible with non-spark as you said. Original issue is sorted though, thank you so much for your help!
